Modify rule S3020: LaYC format (#2822)
SonarSource/rspec: rules/S3020/java/rule.adoc

== Why is this an issue?

The `Collection.toArray()` method returns an `Object[]` when no arguments are provided to it.

> This can lead to a `ClassCastException` at runtime if you try to cast the returned array to an array of a specific class.

Reviewer: ```suggestion
This can lead to a `ClassCastException` at runtime if you try to cast the returned array to an array of a specific type.
```

> Instead, you should use this method by also providing an array of the desired type as the parameter.

Reviewer: ```suggestion
Instead, use this method by providing an array of the desired type as the argument.
```

> Instead, you should use this method by also providing an array of the desired type as the parameter.

Reviewer: Maybe add a sentence about the fact that the array should have length 0?

=== Code examples

==== Noncompliant code example

[source,java,diff-id=1,diff-type=noncompliant]
----
public String [] getStringArray(List<String> strings) {
  return (String []) strings.toArray();  // Noncompliant, a ClassCastException will be thrown here
}
----

==== Compliant solution

[source,java,diff-id=1,diff-type=compliant]
----
public String [] getStringArray(List<String> strings) {

>   return strings.toArray(new String[0]);

Reviewer: Do we need a `// Compliant` comment here?

}
----

== Resources

> * https://docs.oracle.com/javase/8/docs/api/java/util/Collection.html#toArray-T:A-[Collection.toArray()]

Reviewer: The source is missing (see https://github.com/SonarSource/rspec/blob/master/docs/link_formatting.adoc)
